orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -0,0 +1,501 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/logging/logging.h"
+#include "core/common/logging/sinks/clog_sink.h"
+#include "core/framework/tensorprotoutils.h"
+#include "core/graph/graph_viewer.h"
+#include "core/graph/model.h"
+#include "core/platform/path_lib.h"
+#include "core/platform/env.h"
+#include "core/providers/cpu/cpu_execution_provider.h"
+#include "core/util/protobuf_parsing_utils.h"
+#include "onnx/defs/tensor_proto_util.h"
+#include "orttraining/core/framework/checkpointing.h"
+#include "orttraining/core/framework/protobuf_message_sequence.h"
+#include "orttraining/training_api/checkpoint.h"
+
+namespace onnxruntime {
+namespace training {
+namespace api_test {
+
+namespace {
+
+constexpr const char* k_tensor_proto_file_name = ORT_TSTR("tensors.pbseq");
+constexpr const char* k_external_data_file_name = ORT_TSTR("tensors.bin");
+constexpr const char* k_tensor_proto_properties_file_name = ORT_TSTR("properties.pbseq");
+
+constexpr const char* k_param_root_prefix = "param";
+constexpr const char* k_optimizer_root_prefix = "optim";
+constexpr const char* k_property_root_prefix = "custom";
+constexpr const char* k_name_seperator = "_";
+
+PathString GetTensorProtoFilePath(const PathString& checkpoint_directory, const std::string& filename_prefix) {
+  return ConcatPathComponent<PathChar>(checkpoint_directory, ORT_TSTR(filename_prefix + k_name_seperator) + k_tensor_proto_file_name);
+}
+
+PathString GetExternalDataFilePath(const PathString& checkpoint_directory, const std::string& filename_prefix) {
+  return ConcatPathComponent<PathChar>(checkpoint_directory, ORT_TSTR(filename_prefix + k_name_seperator) + k_external_data_file_name);
+}
+
+PathString GetTensorProtoPropertiesFilePath(const PathString& checkpoint_directory, const std::string& filename_prefix) {
+  return ConcatPathComponent<PathChar>(checkpoint_directory, ORT_TSTR(filename_prefix + k_name_seperator) + k_tensor_proto_properties_file_name);
+}
+
+std::string StringConcat(const std::string& s_a, const std::string& s_b, const std::string& del = k_name_seperator) {
+  return s_a + del + s_b;
+}
+
+void StringSplit(const std::string& s, std::vector<std::string>& results, const std::string& del = k_name_seperator) {
+  int start = 0;
```

Comment:
shall we add an assert that the string is non-empty. not sure what happens when we do `s.substr(0, -1)` in case of empty strings.

Reply:
fixed. Thanks.